lowfi_simulation/src/main/native/include/MotorModel/LinearMotorModelSimulation.h

```diff
@@ -0,0 +1,40 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include "MotorModel/MotorModelSimulation.h"
+
+namespace frc {
+namespace sim {
+namespace lowfi {
+
+class LinearMotorModelSimulation : public MotorModelSimulation {
```

Comment:
This class impl makes sense considering this is for lowfi simulation, but LinearMotorModel is probably too generic a name. Any ideal DC brushed motor model is going to be linear, so why this specific one? Is the intent to provide a default impl for more accurate models to override?

Reply:
I don't know much about brushed motors, do they really have a linear response under load?  If I pump 6 volts into a drivetrain geared for 16 fps it can't possibly (at least almost) instantaneously get to 8 fps.  If that is the case I understand why people have been clamoring for getting this added to the supported motor list.

I could rename it, in snobot sim it is called "SimpleMotorModel" but I thought that went too far in the other direction

Reply:
You're confusing a linear step response (of position, velocity, etc.) with the concept of a linear system. A system is linear if the differential equations describing its dynamics consist only of linear operators. Linear operators are things like constant gain multiplications, derivatives, and integrals. You can define reasonably accurate linear models for pretty much everything you'll see in FRC.

The states for a first order model of a DC brushed motor are angular velocity and angle, the output is either angular velocity or angle depending on what your encoder returns, and the input is voltage. You can add more to the model depending on what's around it (like a spinning mass or a carriage in the case of an elevator). The main thing "LinearMotorModelSimulation" is missing is acceleration dynamics (i.e., how the applied voltage changes the velocity over time). This can be determined via the datasheet for your motor. If you're interested, see chapter 12 of https://file.tavsys.net/control/state-space-guide.pdf for how to do this.

You could easily include acceleration dynamics here with a struct or something of motor parameters with the default being a CIM motor or something.

Reply:
Whether to use the name SimpleMotorModel or not depends on how simple it really is. If it uses a non-realistic model internally for just the purposes of testing the code's logic in unit tests, then a name like SimpleMotorModel or MockMotorModel seem appropriate.

Reply:
This guy is supposed to be a super simple "I don't know how motors work, but I want to make sure my auto stops after 5 feet" sort of thing.  Truly linear response with no real word dynamics like back emf or current or acceleration or anything like that.  Based on [this](https://github.com/pjreiniger/SnobotSim/blob/master/snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/motor_sim/SimpleMotorSimulator.java).  SnobotSim also supports better models, like this [static load](https://github.com/pjreiniger/SnobotSim/blob/master/snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/motor_sim/StaticLoadDcMotorSim.java) one

Reply:
Teams shouldn't be using this model to "make sure their auto stops after 5 feet" because it's completely inaccurate. Any autos they write against this model for that purpose like "drive forward for X seconds" won't work for the real robot. It's useful for testing state machine logic and nothing more.

Why is a "truly linear response" (with respect to position over time) desirable in the first place? Having a motor with acceleration included seems like a strictly better alternative. I still think LinearMotorModel is an abuse of the word linear.

Reply:
I'll change the name, I agree with that.

For the first couple years of SnobotSim we only had the "simple model".  I think up until recently RobotPy did the same thing.  Doing it for a drivetrain is less than ideal, but a student lead team might have trouble setting up a better model.  I would argue that a poorly set up model would be more inaccurate than a solid guess at a max speed model.

I do agree that a drivetrain was a bad example.  That is probably the thing that people want the most accurate simulation for, and (from experience) you cannot try to tune a motion profile with that.  A couple years ago we had an elevator  that had top and bottom limit switches, so it was really useful for my kids to add a height simulator hook and test their "if it is at the bottom don't go down, if it is at the top don't go up, otherwise follow the joytsick" code